nvmeof-recoverer: add fast recovery logic for fault scenario

cheolho-kang opened the pull request:
- added python script to handle nvme-of device connection and disconnection
- included `opManagerContext` into the `ClusterManager` for job execution
- implemented `StartNvmeoFConnectJob()` to run the nvmeof connection job on the node
- set up job specifications with HostNetwork to ensure the job runs on the node

this change allows the system to connect nvme-oF devices to other nodes, reusing the fabric devices when an osd pod faults.

<!-- Thank you for contributing to Rook! -->

<!-- STEPS TO FOLLOW:
  1. Add a description of the changes (frequently the same as the commit description)
  2. Enter the issue number next to "Resolves #" below (if there is no tracking issue resolved, **remove that section**)
  3. Review our Contributing documentation at https://rook.io/docs/rook/latest/Contributing/development-flow/
  4. Follow the steps in the checklist below, starting with the **Commit Message Formatting**.
-->

<!-- Uncomment this section with the issue number if an issue is being resolved
**Issue resolved by this Pull Request:**
Resolves #
--->

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/Contributing/development-flow/#commit-structure).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/Contributing/development-flow/#submitting-a-pull-request)
- [ ] [Pending release notes](https://github.com/rook/rook/blob/master/PendingReleaseNotes.md) updated with breaking and/or notable changes for the next minor release.
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.

cheolho-kang committed:
nvmeof-recoverer: add job to connect/disconnect nvme-oF devices

- added python script to handle nvme-of device connection and disconnection
- included `opManagerContext` into the `ClusterManager` for job execution
- implemented `StartNvmeoFConnectJob()` to run the nvmeof connection job on the node
- set up job specifications with HostNetwork to ensure the job runs on the node

this change allows the system to connect nvme-oF devices to other nodes, reusing the fabric devices when an osd pod faults.

Signed-off-by: cheolho.kang <cheolho.kang@samsung.com>

cheolho-kang committed:
nvmeof-recoverer: update nvmeofstorage reconciler to create `OSDTransferInfo` config

- added new struct `OSDTransferInfo` to hold the OSD ID, Node, and Fault Domain information for the OSD transfer.
- added a `createOSDTransferConfig()` to update the `OSDTransferInfo` ConfigMap

this struct will be used to request the operator to relocate the OSD to an another node when the current OSD is down.

Signed-off-by: cheolho.kang <cheolho.kang@samsung.com>

cheolho-kang committed:
nvmeof-recoverer: add preparation step for OSD transfer within fabric fault domain

this commit adds pre-step to prepare for transferring OSDs that need to be moved to a new node within the same fabric fault domain.
- added `transferOSD` field to Cluster struct, to be used for future transfer requests to rook-osd-prepare
- implemented `transferOSDForFabricDomain()` to handle identifying `transferOSD` and deleting failed an OSD
- created `getOSDTransferInfo()` to fetch OSD transfer information from a configmap

these changes set up the groundwork for future PR of OSD transfer requests to rook-osd-prepare.

Signed-off-by: cheolho.kang <cheolho.kang@samsung.com>

cheolho-kang committed:
nvmeof-recoverer: add flags and env vars for preparing osd transfer

this commit introduces new flags and environment variables to prepare for transferring osds within the fabric fault domain.
- added `transferOSDID` and `nvmeOfFaultDomain` flags and env vars to specify the osd to be transferred and its fault domain
- updated `provisionOSDContainer()` to pass these environment variables to the `rook-osd-prepare` job
- these env vars are used by the `rook-operator` to pass flags to the `rook-osd-prepare` job

these changes are necessary for the `rook-osd-prepare` job to find the device for OSD within a fabric fault domain and update the CRUSH location accordingly.

Signed-off-by: cheolho.kang <cheolho.kang@samsung.com>

cheolho-kang committed:
nvmeof-recoverer: bugfix to filter out undesired devices during provisioning

this commit introduces a filter to exclude devices not listed in the desired devices during osd provisioning.
this prevents unintended osds from being configured if a device is not listed in the cephcluster CR.
- add FilterOutUndesiredDevices() to filter out devices not specified in cephcluster CR
- integrate FilterOutUndesiredDevices() into the Provision() to ensure only specified devices are used
- filtering applies when devices are specified in path form (e.g., /dev/nvme0n1)

this fix addresses an issue where unintended osds could be configured, ensuring that only explicitly listed devices are used for osd.

Signed-off-by: cheolho.kang <cheolho.kang@samsung.com>

myoungwon commented on pkg/operator/ceph/nvmeof_recoverer/clustermanager/handler.go:
```diff
@@ -3,28 +3,95 @@ package clustermanager
+		if new_devices:
+			result = '\n'.join(new_devices)
+			print('SUCCESS:', result)
+	except subprocess.CalledProcessError as e:
```
nvme 커맨드가 실패할시 process 의 return 값이 정상은 아닌지 확인 필요

cheolho-kang replied:
except에 걸렸을지라도 성공한 케이스가 있는지 체크해보는 것을 말씀하신걸까요?

myoungwon replied:
넵,

cheolho-kang replied:
넵 반영하였습니다

myoungwon commented on pkg/operator/ceph/nvmeof_recoverer/clustermanager/handler.go:
```diff
@@ -112,3 +179,77 @@ func (cm *ClusterManager) GetNextAttachableHost(currentHost string) string {
+							Command: []string{
+								"python3",
+								"-c",
+								fmt.Sprintf(nvmeofToolCode, string(mode), address, port, subnqn),
```
별개의 스크립트로 만드는건? 추후 생각해볼 문제

cheolho-kang replied:
넵 이부분은 TODO로 넣고 고민해보겠습니다.
별개의 스크립트로 만들었던 이유는 Pod가 뜨면서 수행할 동작을 명시할 때 rook 이 사용하는 방식과 유사하게 하기 위함이었습니다.
(e.g., pkg/operator/ceph/cluster/osd/spec.go 에서 `cephOSDStart`, `activateOSDOnNodeCode`와 같이 OSD Pod를 띄울때 사용할 initContainer들의 동작을 python과 bash command로 만들어서 사용)

myoungwon replied:
resolved 표시는 제가

myoungwon commented on pkg/operator/ceph/nvmeof_recoverer/clustermanager/handler.go:
```diff
@@ -112,3 +179,77 @@ func (cm *ClusterManager) GetNextAttachableHost(currentHost string) string {
+		return "", err
+	}
+
+	var output string
```
TODO 추가 (나중에 더 좋은 확인 방법 필요)

cheolho-kang replied:
넵 TODO 추가하여 추후 fabric device connect 결과의 성공 및 추가된 디바이스의 경로를 가져오는 방법을 개선할 방안 고민해보겠습니다

myoungwon commented on pkg/operator/ceph/nvmeof_recoverer/nvmeofstorage/controller.go:
```diff
@@ -190,12 +191,26 @@ func (r *ReconcileNvmeOfStorage) reconcile(context context.Context, request reco
+					return reconcile.Result{}, err
+				}
+
+				newDevice := extractSuccessMessage(output)
```
공용함수가 아니면 함수화 시킬 필요는 없을듯

cheolho-kang replied:
아하 넵, 이부분은 다시 코드로 넣겠습니다

myoungwon commented on pkg/operator/ceph/cluster/osd/osd.go:
```diff
@@ -824,6 +834,55 @@ func (c *Cluster) applyUpgradeOSDFunctionality() {
+			logger.Debugf(
+				"osd deployment %q not found. Ignoring since object must be deleted.",
+				deploymentName,
+			)
```
예측된 패스가 아니라 assert 같은 것을 넣어야 할것 같습니다.

cheolho-kang replied:
넵 이부분은 단순 에러 로깅이 아닌 assert로 넣겠습니다. (panic(err) 사용)

myoungwon commented on pkg/operator/ceph/cluster/osd/osd.go:
```diff
@@ -824,6 +834,55 @@ func (c *Cluster) applyUpgradeOSDFunctionality() {
+		OSDTransferConfigName, c.clusterInfo.Namespace, delOpts,
+	)
+	if err != nil {
+		return errors.Wrapf(err, "failed to delete the %q configmap", OSDReplaceConfigName)
```
assert 필요

cheolho-kang replied:
넵 이부분도 assert로 넣겠습니다.

myoungwon commented on pkg/operator/ceph/cluster/osd/provision_spec.go:
```diff
@@ -305,6 +305,22 @@ func (c *Cluster) provisionOSDContainer(osdProps osdProperties, copyBinariesMoun
+	if c.transferOSD != nil {
+		if !osdProps.onPVC() {
+			if c.transferOSD.Node == osdProps.crushHostname {
+				envVars = append(
```
기존에 있던 config 체크 하고 append 해야 하지 않을까요?

cheolho-kang replied:
아! 좋은 지적 감사합니다.
저도 그부분 고민해봤었는데요. `envVars`가 `provisionOSDContainer()` 이 함수 내에서만 만들어지고 반환되는데, transfer 관련 env는 여기서만 append하기 때문에 체크는 하지 않아도 될 것 같았습니다. 확인부탁드리며 의견 주시면 반영하여 수정하겠습니다!

myoungwon replied:
what is the role of envVars and how these lines of the codes work?

cheolho-kang replied:
Through this code, envVars are added as environment variables to the `osd-prepare` container. When the provision job runs, the `SetFlagsFromEnv()` converts this `envVars` to `flags`. This process sets the `transferOSDID` and `nvmeOfFaultDomain` flags, which are used to change the CRUSH Location to the `fabric domain`.

myoungwon commented on pkg/daemon/ceph/osd/daemon.go:
```diff
@@ -226,6 +226,11 @@ func Provision(context *clusterd.Context, agent *OsdAgent, crushLocation, topolo
+	// FilterOutUnlistedDevices filters out the devices that are not listed in the desired devices
+	// This can happen when the device is not listed in the desired devices
+	// but it's still configured by ceph-volume
+	deviceOSDs = FilterOutUndesiredDevices(deviceOSDs, agent.devices)
```
이 커밋의 목적이 뭘까요?

cheolho-kang replied:
아 기존 rook 코드의 버그픽스가 목적입니다.

- 기존: 현재 서버에 attach 된 device 중 OSD Device를 모두 deviceOSDs로 사용
- 변경: agent.devices에 명시된 디바이스만 deviceOSDs로 사용하도록 수정
- 이유: clusterCeph CR에는 하기와 같이 두개의 디바이스만 쓰도록 명시하였는데
```
      - name: "qemu2"
        devices:
          - name: "/dev/nvme0n1"
          - name: "/dev/nvme1n1"
```
만약 /dev/nvme2n1과 같은 그 외의 디바이스가 OSD device일 경우 (이전 서버에서 사용하던 device를 임의로 attach 시켰을 경우) 관리자 의도와 다르게 nvme2n1에 대한 OSD도 해당 노드에 생성됨.

따라서 clusterCeph CR 기준으로만 OSD가 뜰 수 있도록 필터링 하기 위함입니다.

myoungwon replied:
Please reorganize the commits to locate this commit at the bottom between the commits.

cheolho-kang replied:
Ok, I will reorganize the commits to position at the bottom of the commits in this PR.

myoungwon commented on cmd/rook/ceph/osd.go:
```diff
@@ -281,7 +281,7 @@ func prepareOSD(cmd *cobra.Command, args []string) error {
 	}
 
-	err = osddaemon.Provision(context, agent, crushLocation, topologyAffinity, deviceFilter, metaDevice)
+	err = osddaemon.Provision(context, agent, crushLocation, topologyAffinity, deviceFilter, metaDevice, nvmeOfFaultDomain, transferOSDID)
```
어쩔수 없긴 한데 함수에 펑션 매개변수 추가할 때는 최대한 안하게 해야 합니다.

cheolho-kang replied:
넵, 피드백 감사합니다. 추후 매개변수 추가 없이 반영할 방법 계속 고민해보겠습니다

myoungwon commented on pkg/operator/ceph/nvmeof_recoverer/nvmeofstorage/controller.go:
```diff
@@ -269,6 +270,40 @@ func (r *ReconcileNvmeOfStorage) fetchNvmeOfStorage(nvmeOfStorage *cephv1.NvmeOf
+	for i, node := range cephCluster.Spec.Storage.Nodes {
+		if node.Name == newNode {
+			newDevice := cephv1.Device{Name: newDevicePath}
+			cephCluster.Spec.Storage.Nodes[i].Devices = append(node.Devices, newDevice)
```
cephCluster.Spec.Storage.Nodes[i].Devices  <-- 이게 뭘까요?

cheolho-kang replied:
cephCluster CR의 Spec.Storage.Nodes 부분을 for loop로 순회하면서 찾은 newNode (=새로 OSD를 붙일 node)의  Devices 리스트를 뜻합니다. 새로 붙인 fabric 디바이스를 cluster CR에 추가시키기 위해 이와같이 하였습니다.

아래는 예시입니다.
```
spec:
  storage:
    nodes:
      - name: "qemu1"
        devices:
          - name: "/dev/nvme0n1"
          - name: "/dev/nvme1n1"
      - name: "qemu2"
         ...
      - name: "qemu3"
         ...
```

myoungwon replied:
check if the `newDevice `is alreay in the `devices`

cheolho-kang replied:
Okay, I will add a check code for the existence of `newDevice`, and since the case that `newDevice` already exists was not considered, I'm going to modify the code to assert fail.

myoungwon commented on pkg/operator/ceph/nvmeof_recoverer/nvmeofstorage/controller.go:
```diff
@@ -269,6 +270,40 @@ func (r *ReconcileNvmeOfStorage) fetchNvmeOfStorage(nvmeOfStorage *cephv1.NvmeOf
+	}
+
+	// Remove the device that is used for the failed OSD from cephCluster CR
+	for i, node := range cephCluster.Spec.Storage.Nodes {
```
remove?

cheolho-kang replied:
아, 코드 개선하면서 남겨진 이전 코멘트였습니다.. 수정하겠습니다

cheolho-kang reviewed (commented):
리뷰 감사합니다. 수정 및 답변 남겼습니다

myoungwon commented on pkg/operator/ceph/nvmeof_recoverer/nvmeofstorage/controller.go:
```diff
@@ -190,12 +191,28 @@ func (r *ReconcileNvmeOfStorage) reconcile(context context.Context, request reco
 				logger.Debugf("next host found: %s", nextHostName)
-				// TODO: Add create and run job for nvme-of device switch to the next host
-				// Placeholder for the job creation and execution
+
```
Before allocating nvme-oF storage to the new host, I think we need to check if the old location is removed correctly, if not, we should umount the original path if you have time to reconsider this.

cheolho-kang replied:
That's a good feedback. Since we need to disconnect after the `OSD Delete` is complete.
I will add this codes after moving the `transfer-related code` from `osd.go`.

myoungwon commented on pkg/operator/ceph/cluster/osd/osd.go:
```diff
@@ -221,6 +222,15 @@ func (c *Cluster) Start() error {
 		return errors.Wrapf(err, "failed to replace an OSD that needs migration to new backend store in namespace %q", namespace)
 	}
 
+	// identify OSDs within the fabric fault domain that need to be transferred to a new node
```
Why do we need this portion? The source host is already deleted before here, isn't it?

myoungwon replied:
Plus, we must avoid making any changes on the original rook.

cheolho-kang replied:
Okay I will change. (7f3837a0a2396aad818041)

The source host was not deleted until this portion.
In the `nvmeofstorage` controller, we only updated the `cluster CR` and marked it as needing to be transferred to new host.
So source host deletion happened here. This approach was taken to sequentially handle multiple OSD compute node failures via rook-operator. 

However, to avoid changes on the original rook, it seems necessary to move this code to the `controller`. 
Based on your feedback, I will move this code to the `controller`, and verify concurrency handling issues later using UT.

myoungwon replied:
> In the `nvmeofstorage` controller, we only updated the `cluster CR` and marked it as needing to be transferred to new host.
> So source host deletion happened here. This approach was taken to sequentially handle multiple OSD compute node failures via rook-operator.

I don't follow this, can you clarify what you explained? Why this codes should be added here instead of nvmeofstroage controller.

myoungwon commented on pkg/daemon/ceph/osd/daemon.go:
```diff
@@ -161,7 +161,7 @@ func configRawDevice(name string, context *clusterd.Context) (*sys.LocalDisk, er
 
 // Provision provisions an OSD
-func Provision(context *clusterd.Context, agent *OsdAgent, crushLocation, topologyAffinity, deviceFilter, metaDevice string) error {
+func Provision(context *clusterd.Context, agent *OsdAgent, crushLocation, topologyAffinity, deviceFilter, metaDevice, nvmeOfFaultDomain string, transferOSDID int) error {
```
This is probably not acceptable when contributing this functionaliy to the upstream. You need to find a better way to implement this by adding the codes to right place (TODO).

cheolho-kang replied:
Okay,

myoungwon commented on pkg/operator/ceph/nvmeof_recoverer/nvmeofstorage/controller.go:
```diff
@@ -271,6 +272,40 @@ func (r *ReconcileNvmeOfStorage) fetchNvmeOfStorage(nvmeOfStorage *cephv1.NvmeOf
+			cephCluster.Spec.Storage.Nodes[i].Devices = append(node.Devices, newDevice)
+		}
+	}
+	_, err = r.context.RookClientset.CephV1().CephClusters(namespace).Update(context.TODO(), cephCluster, metav1.UpdateOptions{})
```
This intends to update the new CR to the rook?

cheolho-kang replied:
Yes, exactly

cheolho-kang committed:
nvmeof-recoverer: update Provision() to handle OSD transfer with NVMe-oF fault domain

this commit updates the Provision() to support transferring OSDs with NVMe-oF fault domains.
- added nvmeOfFaultDomain and transferOSDID parameters to the Provision()
- updated Provision() to adjust the CRUSH location when transferring a device in a fabric fault domain to a new node

these changes ensure that an OSD previously used on another node can be created to a new node without triggering Ceph's rebalancing

Signed-off-by: cheolho.kang <cheolho.kang@samsung.com>

cheolho-kang committed:
nvmeof-recoverer: update nvmeofstorage reconciler to modify `cephcluster` CR during OSD transfer

this commit updates the cephcluster CR in the nvmeofstorage reconciler during the OSD transfer process to accurately reflect the current state of devices and nodes.
- modified `types.go` to specify the cephCluster name in the `nvmeofstorage` CRD
- implemented `updateCephClusterDevices()` to update the `cephcluster` CR with the new device paths and nodes
- updated the `reconcile()` to reflect the cluster changes in the `cephcluster` CR
- 'deploy/' and 'Documentation/' directories are auto-generated code

these changes ensure that the cephcluster CR accurately reflects the current state of devices and nodes during OSD transfers.

Signed-off-by: cheolho.kang <cheolho.kang@samsung.com>

cheolho-kang committed:
nvmeof-recoverer: update reconcile to remove crashloopbackoff osd

this commit enhances the `nvmeofstorage` controller by adding logic to handle osd deployments that are in crashloopbackoff state.
- added code to delete an osd deployment in crashloopbackoff state using `k8sutil.DeleteDeployment()`
- added disconnection of the fabric device used by the osd

cheolho-kang commented on pkg/operator/ceph/cluster/osd/transfer.go:
```diff
@@ -27,3 +35,29 @@ type OSDTransferInfo struct {
+
+// getOSDTransferInfo returns an existing OSD info that needs to be transfered to a new node
+func (c *Cluster) getOSDTransferInfo() (*OSDTransferInfo, error) {
+	cm, err := c.context.Clientset.CoreV1().ConfigMaps(c.clusterInfo.Namespace).Get(
```
configmap을 가져와서 c.transferOSD를 세팅하는데 사용되는 부분입니다.
이 변수는 `rook-osd-prepare` 에게 이동시킨 OSD의 crush location을 fabric failure domain으로 설정하기 위해 쓰입니다.

cheolho-kang commented on pkg/operator/ceph/cluster/osd/osd.go:
```diff
@@ -942,6 +952,34 @@ func (c *Cluster) getOSDStoreStatus() (*cephv1.OSDStatus, error) {
 }
 
+// checkNeedToTransferOSD checks config map if there is an OSD using fabric failure domain that needs to be transferred to a new host
+func (c *Cluster) checkNeedToTransferOSD() error {
```
`nvmeofstorage controller`가 생성한 configmap을 가져와서 c.transferOSD를 세팅하는데 사용하는 부분만 남겼습니다.
`rook-osd-prepare` 변수를 전달하기 위해선 operator 코드에서만 가능하여 남겨놨고,
해당 변수는 이동시킨 OSD의 crush location을 fabric failure domain으로 설정하기 위해 쓰입니다.

[변수 사용처](https://github.com/cheolho-kang/rook/pull/4/commits/7f3837a0a2396aad8180415cbb4a6032bac6a236#r1642145409)

myoungwon commented on pkg/operator/ceph/nvmeof_recoverer/nvmeofstorage/controller.go:
```diff
@@ -181,21 +182,32 @@ func (r *ReconcileNvmeOfStorage) reconcile(context context.Context, request reco
+		// Connect the device to the next host for fast recovery
+		var newDevice string
+		newDevice, err = r.clustermanager.StartNvmeoFConnectJob(cm.NvmeofConnect, nextHostName,
+			r.nvmeOfStorage.Spec.IP, strconv.Itoa(targetOSDInfo.Port), targetOSDInfo.SubNQN)
```
`r.nvmeOfStorage.Spec.IP `does not seem necessary here. Probably move it to the `StartNvmeoFConnectJob

myoungwon commented on pkg/operator/ceph/nvmeof_recoverer/clustermanager/handler.go:
```diff
@@ -112,3 +183,86 @@ func (cm *ClusterManager) GetNextAttachableHost(currentHost string) string {
 	return ""
 }
+
+func (cm *ClusterManager) StartNvmeoFConnectJob(mode NvmeofMode, targetHost, address, port, subnqn string) (string, error) {
```
StartNvmeoFConnectJob --> ConnectNvmeoF ?

myoungwon commented on pkg/operator/ceph/nvmeof_recoverer/nvmeofstorage/controller.go:
```diff
@@ -286,6 +289,48 @@ func (r *ReconcileNvmeOfStorage) fetchNvmeOfStorage(nvmeOfStorage *cephv1.NvmeOf
+			}
+			cephCluster.Spec.Storage.Nodes[i].Devices = updatedDevices
+		}
+	}
```
Can we merge two for codes into one ?
